### PR TITLE
PLT-648 Enhance performance of Basics

### PR DESCRIPTION
**Pull request by jodastephen**
- Faster internals of Currency
- Faster date manipulation
- Faster tenor/period adjustment
- Faster schedule generation

### CODE REVIEW

File: modules/basics/src/main/java/com/opengamma/basics/date/LocalDateUtils.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+package com.opengamma.basics.date;
+
+import java.time.LocalDate;
+
+/**
+ * Utilities for working with {@code LocalDate}.
+ */
+final class LocalDateUtils {
+
+  // First day-of-month minus one for a standard year
```
Comment (cjkent) [inline modules/basics/src/main/java/com/opengamma/basics/date/LocalDateUtils.java]: Add a comment explaining what the extra zero is doing in position zero.

File: modules/basics/src/main/java/com/opengamma/basics/date/LocalDateUtils.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+   * 
+   * @param date  the date to add to
+   * @return the new date
+   */
+  static LocalDate plusDays(LocalDate date, int daysToAdd) {
+    if (daysToAdd == 0) {
+      return date;
+    }
+    long dom = date.getDayOfMonth() + daysToAdd;
+    if (dom > 0 && dom <= 59) { // 59th Jan is 28th Feb, 59th Feb is 31st Mar
```
Comment (cjkent) [inline modules/basics/src/main/java/com/opengamma/basics/date/LocalDateUtils.java]: A bit of explanation of what's happening here.

Comment (cjkent) [inline modules/basics/src/main/java/com/opengamma/basics/date/LocalDateUtils.java]: Is there any need for a fast path for negative ``daysToAdd``? Or is it mostly used for moving forwards?

Comment (cjkent) [inline modules/basics/src/main/java/com/opengamma/basics/date/LocalDateUtils.java]: Just read the code more carefully. So moving backwards out of the current month is slower? Is that right? Is there any need to optimise that?

File: modules/basics/src/main/java/com/opengamma/basics/date/PeriodAdjustment.java [hunk 3]
```diff
@@ -162,7 +162,7 @@ private void validate() {
   public LocalDate adjust(LocalDate date) {
     ArgChecker.notNull(date, "date");
     LocalDate unadjusted = additionConvention.adjust(date, period, adjustment.getCalendar());
-    return unadjusted.with(adjustment);
+    return adjustment.adjust(unadjusted);
```
Comment (cjkent) [inline modules/basics/src/main/java/com/opengamma/basics/date/PeriodAdjustment.java]: What's going on here?

File: modules/basics/src/main/java/com/opengamma/basics/schedule/PeriodicSchedule.java [hunk 4]
```diff
@@ -349,12 +348,22 @@ private void validate() {
...
+      // for performance, handle silly errors using exceptions
+      for (int i = 0; i < unadj.size() - 1; i++) {
+        periods.add(SchedulePeriod.of(adj.get(i), adj.get(i + 1), unadj.get(i), unadj.get(i + 1)));
+      }
+    } catch (IllegalArgumentException ex) {
+      // check dates to throw a better exception for duplicate dates in schedule
+      createUnadjustedDates();
+      createAdjustedDates();
+      // unknown exception
+      throw (ScheduleException) new ScheduleException(
```
Comment (cjkent) [inline modules/basics/src/main/java/com/opengamma/basics/schedule/PeriodicSchedule.java]: This is a very surprising piece of code. It's obvious what it does once you read the Javadocs for ``initCause``, but would it be clearer to do it in a slightly more conventional way:

    ScheduleException scheduleException = new ScheduleException(...);
    scheduleException.initCause(ex);
    throw scheduleException;

File: modules/basics/src/main/java/com/opengamma/basics/date/LocalDateUtils.java (hunk 2 above)
Comment (jodastephen) [inline modules/basics/src/main/java/com/opengamma/basics/date/LocalDateUtils.java]: Its all trade offs, moving backward within a month is fast, moving backward to the previous month is slow.

File: modules/basics/src/main/java/com/opengamma/basics/date/PeriodAdjustment.java (hunk 3 above)
Comment (jodastephen) [inline modules/basics/src/main/java/com/opengamma/basics/date/PeriodAdjustment.java]: New way round calls a specific method typed to accept LocalDate, whereas the previous way had to check the object multiple times and call logic via an interface. Faster.

**Pull request merged**